### PR TITLE
Some requests in polyhedral geometry

### LINKED ISSUE #3748: Casting `PolyhedralFan` to `PolyhedralComplex`
It happens now and then that I want to convert a polyhedral fan to a polyhedral complex in Oscar. There are already versions of `polyhedral_complex` that do this for polyhedra and tropical varieties, so perhaps it could make sense to have one also for fans? 

I'm currently using the following function locally (thanks for @lkastner for helping me with this over on Slack!). Perhaps something along these lines should be in Oscar? 

```
function polyhedral_complex_from_fan(F::PolyhedralFan)
    K = F.parent_field
    rml, L = rays_modulo_lineality(F)
    rays_and_vertex = vcat(matrix(K, rml), zero_matrix(ZZ, 1, ambient_dim(F)))
    cells = maximal_cones(IncidenceMatrix, F)
    new_cells = IncidenceMatrix(nrows(cells), ncols(cells)+1);
    new_cells[1:nrows(cells), 1:ncols(cells)] = cells;
    for i in 1:nrows(cells) 
        new_cells[i, ncols(cells)+1] = 1
    end
    return polyhedral_complex(new_cells, rays_and_vertex, collect(1:length(rml)))
end
```

### PR DESCRIPTION
**Pull request by lkastner**
- Implement `polyhedron(Cone)`
- Implement `polyhedral_complex(PolyhedralFan)`
- Add warning for subdivision of points #3785

### CODE REVIEW

**Comment by lkastner** [comment]
cc @YueRen @oskarhenriksson @benlorenz

**Review by YueRen — APPROVED** [review]
Looks good, thanks a lot!

File: src/PolyhedralGeometry/PolyhedralComplex/constructors.jl [hunk 1]
```diff
@@ -170,6 +170,38 @@ end
...
+      inprop = Polymake.give(pmo_in, prop)
+      outprop = Polymake.IncidenceMatrix(nrows(inprop), ncols(inprop)+1)
+      outprop[1:end, 2:end] = inprop
+      for i in 1:nrows(inprop)
+        outprop[i,1] = 1
+      end
+      Polymake.take(pmo_out, prop, outprop)
+    end
+  end
+  return PolyhedralComplex{T}(pmo_out, F.parent_field)
```
Comment (alexej-jordan) [inline src/PolyhedralGeometry/PolyhedralComplex/constructors.jl]:
```suggestion
  return PolyhedralComplex{T}(pmo_out, coefficient_field(F))
```
We should avoid directly accessing fields of a struct and use the specific functions instead.

File: src/PolyhedralGeometry/Polyhedron/constructors.jl [hunk 2]
```diff
@@ -290,6 +294,29 @@ convex_hull(
...
+    if Polymake.exists(pmo_in, prop)
+      Polymake.take(pmo_out, prop, embed_at_height_one(Polymake.give(pmo_in, prop), true))
+    end
+  end
+  for prop in ["INPUT_LINEALITY", "LINEALITY_SPACE", "EQUATIONS", "LINEAR_SPAN"]
+    if Polymake.exists(pmo_in, prop)
+      Polymake.take(pmo_out, prop, embed_at_height_one(Polymake.give(pmo_in, prop), false))
+    end
+  end
+  return Polyhedron{T}(pmo_out, C.parent_field)
```
Comment (alexej-jordan) [inline src/PolyhedralGeometry/Polyhedron/constructors.jl]:
```suggestion
  return Polyhedron{T}(pmo_out, coefficient_field(C))
```
As above

File: src/PolyhedralGeometry/PolyhedralComplex/constructors.jl [hunk 3]
```diff
@@ -170,6 +170,49 @@ end
...
 
+@doc raw"""
+    polyhedral_complex(F::PolyhedralFan)
+
+Turn a polyhedral fan into a polyhedral complex.
+"""
+function polyhedral_complex(F::PolyhedralFan{T}) where {T<:scalar_types}
+  pmo_in = pm_object(F)
+  pmo_out = Polymake.fan.PolyhedralComplex{_scalar_type_to_polymake(T)}()
+  for prop in ["RAYS", "INPUT_RAYS", "FACET_NORMALS"]
```
Comment (fingolfin) [inline src/PolyhedralGeometry/PolyhedralComplex/constructors.jl]: Looking at this I wondered "hmm, how many allocations will this make? Can it reuse the strings? What about the vector?"

Turns out if you change this from a vector to a tuple, you can avoid all allocations in that line. I.e.:
```suggestion
  for prop in ("RAYS", "INPUT_RAYS", "FACET_NORMALS")
```

To demonstrate the difference:
```julia
julia> function f(s)
         for prop in ["MAXIMAL_CONES", "INPUT_CONES", "MAXIMAL_CONES_FACETS"]
           s == prop && return true
         end
         return false
       end
f (generic function with 1 method)

julia> function g(s)
         for prop in ("MAXIMAL_CONES", "INPUT_CONES", "MAXIMAL_CONES_FACETS")
           s == prop && return true
         end
         return false
       end
g (generic function with 1 method)

julia> @btime f("MAXIMAL_CONES")
  25.435 ns (1 allocation: 80 bytes)
true

julia> @btime g("MAXIMAL_CONES")
  1.500 ns (0 allocations: 0 bytes)
true
```
Mind you, either is plenty fast and I assume it won't matter for `polyhedral_complex` which hopefully nobody will call in a tight loop ;-) so this is *not* a change request -- I just wondered about the behaviour of this code myself and thought I'd share my findings for future reference :-)

Comment (lkastner) [reply src/PolyhedralGeometry/PolyhedralComplex/constructors.jl]: Nice, thank you.

Comment (YueRen) [reply src/PolyhedralGeometry/PolyhedralComplex/constructors.jl]: I learned something new in julia today.  So is it generally recommended to use tuples over vectors when (easily) possible?

**Pull request merged**